app/plugins/hmac-authentication.md

```diff
@@ -115,45 +133,189 @@ the order they appear.
+format:
 
-The server and requesting client should be synchronized with NTP and a valid date (using GMT format) should be sent with either the `X-Date` or `Date` header.
+```
+Digest: SHA-256=base64(sha256(<body>))
+```
 
-### HMAC Example
+Note: In order to create digest of request body, the plugin needs to retain
+it in memory, which might cause pressure on the worker's Lua VM when dealing
+with large bodies (several MBs) or during high request concurrency.
 
-For an HMAC signature with `date` and `content-md5` headers, the `Proxy-Authorization` or `Authorization` header and signature would be generated as:
+### Enforcing Headers
 
-```
-Authorization: hmac username="bob", algorithm="hmac-sha1", headers="date content-md5", signature="Base64(HMAC-SHA1(signing string))"
-```
+`config.enforce_headers` can be used to enforce any of the headers to be part 
+of signature creation. Plugin by default doesn't enforce which header need to
+be used for signature creation. The minimum recommended data to sign is the
+`request-line`, `host`, and `date`. A strong signature including all of the
+headers and a `digest` of the body.
 
-The client would compose the signing string as:
+### HMAC Example
 
-```
-date: Fri, 09 Oct 2015 00:00:00 GMT\ncontent-md5: lCMsW4/JJy9vc6HjbraPzw==
-```
+  **Add an API**
+
+  ```bash
+  $ curl -i -X POST http://localhost:8001/apis \
+      -d "name=hmac-test" \
+      -d "hosts=hmac.com" \
+      -d "upstream_url=http://mockbin.com"
+  HTTP/1.1 201 Created
+  ...
+
+  ```
+  
+  **Enable plugin**
+
+  ```bash
+  $ curl -i -X POST http://localhost:8001/apis/hmac-test/plugins \
+      -d "name=hmac-auth" \
+      -d "config.enforce_headers=date, request-line" \
+      -d "config.algorithms=hmac-sha1, hmac-sha256"
+  HTTP/1.1 201 Created
+  ...
+
+  ```
+
+  Here we are enabling `hmac-auth` plugin on API `hmac-test`.
+  `config.enforce_headers` is set to force client to at least use
+  `date` and `request-line` in http signature signature creation. 
+  Also we are setting `config.algorithms` to force client only use
+  `hmac-sha1` or `hmac-sha256` for hashing signing string.
+
+  **Add a consumer**
+
+  ```bash
+  $ curl -i -X POST http://localhost:8001/consumers/ \
+      -d "username=alice"
+  HTTP/1.1 201 Created
+  ...
+
+  ```
+
+  **Add credential for Alice**
+
+  ```bash
+  $ curl -i -X POST http://localhost:8001/consumers/alice/hmac-auth \
+      -d "username=alice123" \
+      -d "secret=secret"
+  HTTP/1.1 201 Created
+  ...
+
+  ```
+
+  **Request to the API**
+
+  ```bash
+  $ curl -i -X GET http://localhost:8000/requests \
+      -H "Host: hmac.com" \
+      -H "Date: Thu, 22 Jun 2017 17:15:21 GMT" \
+      -H 'Authorization: hmac username="alice123", algorithm="hmac-sha256", headers="date request-line", signature="ujWCGHeec9Xd6UD2zlyxiNMCiXnDOWeVFMu5VeRUxtw="'
+  HTTP/1.1 200 OK
+  ...
+  
+  ```
+  
+  For the above request we are composing signing string using `date` and
```

Comment:
*In* the above request, we are composing *the* signing string using *the* `date` and `request-line` headers, ... *using the `hmac-256` algorithm to hash the digest*?